### PR TITLE
Add events to tracking for PMC/PMUI powershell.

### PR DESCRIPTION
**Pull request by erdembayar**
## Fix

Details: [Issue#10069](https://github.com/NuGet/Home/issues/10069)
Child of https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1167790.
Spec for this PR: https://github.com/NuGet/Client.Engineering/pull/553/files

@dominoFire Please give your thought.
I am not trying to answer all questions now or propose all the possible questions/hypothesis without even basic telemetry. Let's just have minimum telemetry for PMC and init1.ps1 scripts and commands run on PMC. Then after looking at that one let's decide if we need more telemetry. Problem is currently we don't have any telemetry for above.

I created this PR for telemetry, once this is merged we're going to have much better picture of PMC and what users do around it.
Main intention for this PR is understand about PMC and how often user does run commands etc.

For example if sum of `point 3` + `point 4` is `3000 `and but `point 2` is just `30` then it's pretty much self explanatory, that means user wanted to open PMC intentionally most of the time, so we can't do anything about it.

1. [ReOpenAtStart](https://github.com/NuGet/NuGet.Client/blob/9f84b339bc8abf1dcdb2e4f975ac2a1a5865c8dd/src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs#L98): One of issue raised by Calvin was PMC was last open tool window then it get automatically loaded when user open VS again.
    This is implicit action, it helps with learn about user behavior and actions.
    With this change we can say exactly how many times this happen and it helps with answer if it's really the issue.
    Now this `PackageManagerConsoleDefaultOpen` (renamed to `ReOpenAtStart`) event is combined to previous event to reduce number of events to emit.

image

2. [PackageManagerConsoleLoadCount](https://github.com/NuGet/NuGet.Client/blob/9f84b339bc8abf1dcdb2e4f975ac2a1a5865c8dd/src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs#L89): Now with this change we can answer how many times PMC was loaded/opened during each  VS session. 
image

3. I already have query to find out how many times  PMC was opened from View>>Other>>PMC.
    This is explicit action, it helps with learn about user behavior and actions.
4. Using above 3 info we can `calculate` how many times below action happens. (`@2-@1-@3 = @4`)
    `Have PMC available in the window toolbar, and loads only when you click on it to focus.`
    This is explicit action, it helps with learn about user behavior and actions.
5. [PowerShellExecuteCommand](https://github.com/NuGet/NuGet.Client/blob/54991a3e730b9f949e9ba34ad1ce879f2c19999d/src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs#L343): With this we can answer how often user actually execute commands on PMC.

image

image

## Testing/Validation

Tests Added: Yes/No  
Reason for not adding tests:  
Validation:

### CODE REVIEW

**Comment by JonDouglas** [comment]
I think these are additions & I believe the powershell execute event will help us understand how often that is happening for the cost of loading powershell.

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs [hunk 1]
```diff
@@ -75,6 +80,26 @@ public void NotifyInitializationCompleted()
...
+                var telemetryEvent = new TelemetryEvent("PackageManagerConsoleLoadCount", new Dictionary<string, object>
+                                {
+                                    { "NugetPMCLoadCount", _windowLoadCount}
+                                });
+                TelemetryActivity.EmitTelemetryEvent(telemetryEvent);
+
+                // Work around to detect if PMC loaded automatically because it was last focused window.
+                if (IsLoaded)
+                {
+                    telemetryEvent = new TelemetryEvent("PackageManagerConsoleDefaultOpen");
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs]: Does this really mean "Default Open"? IsLoaded is just telling you that PMC has been loaded at some point, but do you really know when?

Comment (erdembayar) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs]: Default open: PMC loaded automatically because it was last focused window before we close VS.
Instead of trying to figure out when did PMC was opened (which is hard), we solve problem from backward.
It  just detect if PMC is active/focused window before VS close, so it's analogues to same problem above.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs]: Hmm...what happens when a user installs a package with an install.ps1 and the PMC initializes, and then they open PMC and focus on it? 

Does it send this event?

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs [hunk 2]
```diff
@@ -75,6 +80,26 @@ public void NotifyInitializationCompleted()
 
         public void Dispose()
         {
+            if (!_isTelemetryEmitted)
+            {
+                var telemetryEvent = new TelemetryEvent("PackageManagerConsoleLoadCount", new Dictionary<string, object>
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs]: I think it's better if you think about this in a `1 event, lots of data` way. 

See https://github.com/NuGet/Client.Engineering/issues/319. 

That way analysis will be easier because you only need to query for 1 event.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs [hunk 3]
```diff
@@ -325,6 +327,16 @@ public void Initialize(IConsole console)
 
                                 DefaultProject = null;
 
+                                var telemetryEvent = new TelemetryEvent("PowerShellExecuteCommand", new Dictionary<string, object>
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: It might be helpful to add abstractions for the telemetry events, similar to RestoreCommand.

Comment (nkolev92) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: When is being sent again?
It looks to me that it'll only be sent after init.ps1 runs.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs [hunk 4]
```diff
@@ -484,6 +499,11 @@ private async Task ExecuteInitPs1Async(string installPath, PackageIdentity ident
                             request.BuildInput(),
                             outputResults: true);
 
+                        var telemetryEvent = new TelemetryEvent(eventName: "PowershellScriptExecuted");
+                        telemetryEvent.AddPiiData("id", identity.Id?.ToLowerInvariant() ?? "(empty package id)");
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: What's the value of getting the ids?

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs [hunk 5]
```diff
@@ -528,6 +548,16 @@ public bool Execute(IConsole console, string command, params object[] inputs)
...
             }
 
+            // Command can come here from both PMC and PM UI.
+            if (console is IWpfConsole)
+            {
+                _pmcExecutedCount++;
+            }
+            else
+            {
+                _nonPmcExecutedCount++;
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: Does this happen when an install script runs?

**Review by nkolev92 — COMMENTED** [review]
* Think fewer events.
* Consider creating abstractions.

* Instead of screenshots, it'll nice if you provide text data. It'll make switching back and forth easy and allow us to search for the description correlating the event.

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs (hunk 1 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs]: Yes, it'll be sent. Because when you install package with an install.ps1 then it only initialize Powershell only, not PMC. So if you open PMC later and focus on it then  it's separate event so it'll be sent when user close VS instance. But even you install EF6 with install.ps1 then if you don't open PMC at all then this event wouldn't be sent at all.
Actually PMUI and PMC creates separate instances of `PowerShellHost.cs`.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs (hunk 3 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs:330]: This one is hooked into `close solution event`. So it'll be only sent when user close the VS solution regardless of init.ps1.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs (hunk 4 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: Currently no immediate use. But during last Teams call someone mentioned we're not tracking this events, specially since EF6 one of most used package we need to pay attention to how people use it.
There could be other packages use powershell init.ps1 extensively. I believe we need this information later when we analyze use cases of PMC with powershell.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs (hunk 3 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs:330]: That abstraction is useful for wrapper around method and emit duration. But here only emit once for `solution close event` so I don't see any usefulness, having `using` statement with empty body doesn't make any sense.

Comment (erdembayar) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: > It might be helpful to add abstractions for the telemetry events, similar to RestoreCommand.

That abstraction is useful for wrapper around method and emit duration. But here only emit once for `solution close event` so I don't see any usefulness, having `using` statement with empty body doesn't make any sense.

Comment (erdembayar) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: > When is being sent again?
> It looks to me that it'll only be sent after init.ps1 runs.

This one is hooked into `close solution event`. So it'll be only sent when user close the VS solution regardless of init.ps1 as long as powershell is already loaded( either from PMC or PM UI).

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs (hunk 1 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs]: Ok, so now if I understand, you're trying to write that it WILL BE opened by default when VS is opened next time? In that case, I suggest something that is less confusing without that context. Maybe "WillReopen" or "OpenAtShutdown" or something?

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs (hunk 4 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: nit: PowerShell

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs [hunk 6]
```diff
@@ -325,6 +327,16 @@ public void Initialize(IConsole console)
 
                                 DefaultProject = null;
 
+                                var telemetryEvent = new TelemetryEvent("PowerShellExecuteCommand", new Dictionary<string, object>
+                                {
+                                    { "NugetPMCExecuteCommandCount", _pmcExecutedCount},
+                                    { "NugetNonPMCExecuteCommandCount", _nonPmcExecutedCount},
+                                    { "NugetTotalExecuteCommandCount", _pmcExecutedCount + _nonPmcExecutedCount}
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: nit: NuGet

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs (hunk 5 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: ~~Yes, it does if it's originated from PM UI.~~ 
Update. If you're asking about 'install.ps1' then it doesn't. But it count install action itself.

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs (hunk 1 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs:94]: Actually PMUI and PMC creates separate instances of `PowerShellHost.cs`.

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs (hunk 2 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs]: Ok. I combined 2 events in `ConsoleContainer.xaml.cs` into 1 event.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs (hunk 4 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: Ok. Fixed it.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs (hunk 6 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: Ok. Fixed it.

**Comment by erdembayar** [comment]
> * Think fewer events.
I combined 2 event from `ConsoleContainer.xaml.cs`, but others (`PowerShellExecuteCommand` and `PowerShellScriptExecuted`) remain as it's because it's hard to combine them. I found no existing example of add list into telemetry event. I was considering add list of `PowerShellScriptExecuted into `PowerShellExecuteCommand`, but it's not straight forward.
> * Consider creating abstractions.
I already answered it.
> * Instead of screenshots, it'll nice if you provide text data. It'll make switching back and forth easy and allow us to search for the description correlating the event.
Ok. I added text naming of events I added into description part, next time I'll use more links and texts.

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs (hunk 1 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs]: Ok. How about "ReOpenAtStart"? I believe this one better reflect what it does.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs (hunk 4 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: Removed since there is no planned use for now. Also it doesn't capture all cases.

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs [hunk 7]
```diff
@@ -75,6 +84,27 @@ public void NotifyInitializationCompleted()
 
         public void Dispose()
         {
+            if (!_isTelemetryEmitted)
+            {
+                var telemetryEvent = new TelemetryEvent(PackageManagerConsoleLoadCount, new Dictionary<string, object>
+                                {
+                                    { NuGetPMCLoadCount, _windowLoadCount},
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs:93]: When does `Dispose` get called?  If it gets called when the PMC window is closed, then `_windowLoadCount` will always equal exactly 1.

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs [hunk 8]
```diff
@@ -75,6 +84,27 @@ public void NotifyInitializationCompleted()
...
+                                    { ReOpenAtStart, false}
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs]: What's the purpose of sending a property whose value is always the same?

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs [hunk 9]
```diff
@@ -314,20 +321,37 @@ public void Initialize(IConsole console)
...
-                        // check if PMC console is actually opened, then only hook to solution load/close events.
+                        // check if PMC console is actually opened, then only hook to solution load events.
                         if (console is IWpfConsole)
                         {
                             // Hook up solution events
                             _solutionManager.Value.SolutionOpened += (_, __) => HandleSolutionOpened();
-                            _solutionManager.Value.SolutionClosed += (o, e) =>
+                        }
+
+                        _solutionManager.Value.SolutionClosed += (o, e) =>
```
Comment (zivkan) [inline src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs:337]: I was expecting a meeting where we could discuss the telemetry at a high level, rather than diving straight into the implementation immediately.

NuGet starts the powershell host on first usage, whether that's a packages.config package install that uses install.ps1, or opening PMC. The PowerShell host then runs for the rest of the devenv process, regardless of the solutions.

Therefore, what happens when:
* A customer starts VS, does not load a solution, and uses PMC, closes VS. Will we get any telemetry?
* A customer starts VS, loads a solution, does not use PowerShell, close the solution, open second solution in the same process, opens PMC.
* A customer starts VS, loads a solution, loads PMC, closes PMC, opens a second solution and does not use PMC.

In the last two scenarios, the powershell host is in memory, but if we query telemetry, it will look like the powershell host was only loaded 50% of the time, whereas it was actually loaded in 75% of the solution sessions, and 100% of the VS sessions.

Furthermore, you're adding this telemetry in NuGetConsole's PowerShellHost. Does't this mean that a devenv process that never uses NuGet's powershell host will not generate any telemetry?  How will we determine the percentage of VS sessions that use powershell, without telemetry events that say "feature was not used"?

Your PR says
> I am not trying to answer all questions now or propose all the possible questions/hypothesis without even basic telemetry. Let's just have minimum telemetry for PMC and init1.ps1 scripts and commands run on PMC. Then after looking at that one let's decide if we need more telemetry. Problem is currently we don't have any telemetry for above.

The problem I was trying to state in our previous meetings is that if we don't figure out what questions we want to answer, then the telemetry we add might be useless. If we don't consider the question first, then we implement telemetry, like this PR is proposing, then we have to wait until hotseat inserts, then we have to wait for the version to ship, then we start getting data. If we query the data and find it doesn't help us make decisions, we've wasted weeks or months. This is exactly the problem I had with the package source telemetry I added nearly a year ago, so I know from first-hand experience. Once we started getting data and tried to query it, it didn't give us useful information to make decisions.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs [hunk 10]
```diff
@@ -314,20 +321,37 @@ public void Initialize(IConsole console)
...
+                        {
+                            var telemetryEvent = new TelemetryEvent(PowerShellExecuteCommand, new Dictionary<string, object>
+                                {
+                                    { NuGetPMCExecuteCommandCount, _pmcExecutedCount},
+                                    { NuGetNonPMCExecuteCommandCount, _nonPmcExecutedCount},
+                                    { NuGetTotalExecuteCommandCount, _pmcExecutedCount + _nonPmcExecutedCount},
+                                    { IsIWpfConsole, console is IWpfConsole}
```
Comment (zivkan) [inline src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: Does this mean that when a customer installs a package with install.ps1 and they use PMC in the same session, that we generate two telemetry events? DOes NuGet create two powershell hosts?

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs (hunk 7 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs:93]: It's called only when you close VS instance. Hiding or switching to other tabs doesn't cause it dispose. This number does increase more than 1. It increases every time user unhide or switch to from other tabs.

image

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs (hunk 8 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs]: We do override this value at [line98](https://github.com/NuGet/NuGet.Client/blob/2c252751bfb944bfb1df226e07ba818558a84231/src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs#L98). But having place holder property like this makes writing querying much easier and faster. Otherwise have to write another condition check if this property/column exist which makes query slower.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs (hunk 10 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: Yes, PMC and PMUI creates 2 separate powershell hosts. So it'll be 2 events, and maybe we can make it single event but it involves static variables which I prefer not to do it.

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs (hunk 8 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs]: Doesn't `{ ReOpenAtStart, IsLoaded }` provide the same result?  Or
```cs
var reopenAtStart = IsLoaded;
var telemetryEvent = new TelemetryEvent(/* ... */)
  {
    // other things
    { ReOpenAtStart, reopenAtStart }
  };
```

I acknowledge that my comment is very subjective, but it was a real surprise to me that it looked like a hardcoded value was being added to the telemetry. It surprised me so much, I didn't notice it being overwritten a few lines further down.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs (hunk 10 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: That's very interesting. I didn't know that.

An alternative to using statics, is to raise events in the PowerShell host, and have something running in the "higher level" Visual Studio code listen to the events and aggregate the data. My package source telemetry was problematic with regards to data quality and being able to answer questions, but I think it's the first NuGet telemetry that is aggregated, rather than "emit at source". I used a .NET event, but now I think a ETW event would be better. Listening to ETW events is a bit more difficult than subscribing to .NET events, but it eliminates the need for public APIs (public static event).

Basically, it could work the similarly to System.Net.Http's diagnostics: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/HttpDiagnosticsGuide.md

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs (hunk 7 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs:93]: Does our XAML instance getting disposed at VS close, rather than PMC window close, sound by-design, or like a bug?  To me, it sounds like a bug, so I'm worried that relying on this behaviour for telemetry will start producing bad telemetry if the underlying memory leak is fixed (at least this memory leak seems to only create a single instance)

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs (hunk 9 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs:337]: > Therefore, what happens when:
> 
> * A customer starts VS, does not load a solution, and uses PMC, closes VS. Will we get any telemetry?

We still get [PackageManagerConsoleLoadCount](https://github.com/NuGet/NuGet.Client/blob/9f84b339bc8abf1dcdb2e4f975ac2a1a5865c8dd/src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs#L89), [ReOpenAtStart](https://github.com/NuGet/NuGet.Client/blob/9f84b339bc8abf1dcdb2e4f975ac2a1a5865c8dd/src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs#L98) events this case, and I imagine it hardly ever happen and but we'll know when it happens. But we don't get [PowerShellExecuteCommand](https://github.com/NuGet/NuGet.Client/blob/9f84b339bc8abf1dcdb2e4f975ac2a1a5865c8dd/src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs#L334) event this case since there is no solution.

> * A customer starts VS, loads a solution, does not use PowerShell, close the solution, open second solution in the same process, opens PMC.
> * A customer starts VS, loads a solution, loads PMC, closes PMC, opens a second solution and does not use PMC.

>In the last two scenarios, the powershell host is in memory, but if we query telemetry, it will look like the powershell host was only loaded 50% of the time, whereas it was actually loaded in 75% of the solution sessions, and 100% of the VS sessions.

I'm not aware of use case for how many % of solution session duration had poweshell host is loaded, we're more interested how many times it's loaded without user's intention of use. If `powershell is loaded without users intention of` use less than 0.1% (which I believe) then there is not much to dig more since it's very rare case, but if it happen 5%-10% of all VS open session then above question is more interesting. 

> Furthermore, you're adding this telemetry in NuGetConsole's PowerShellHost. Does't this mean that a devenv process that never uses NuGet's powershell host will not generate any telemetry? How will we determine the percentage of VS sessions that use powershell, without telemetry events that say "feature was not used"?

VS itself already generate ton of telemetry, no need to add another telemetry for `feature was not used`. We can aggregate total number of sessions from Kusto table. Current new telemetry is for `powershell loaded case`. So `powershell %` =   `powershell loaded case`/ `Total sessions`.

>The problem I was trying to state in our previous meetings is that if we don't figure out what questions we want to answer, then the telemetry we add might be useless. If we don't consider the question first, then we implement telemetry, like this PR is proposing, then we have to wait until hotseat inserts, then we have to wait for the version to ship, then we start getting data. If we query the data and find it doesn't help us make decisions, we've wasted weeks or months. This is exactly the problem I had with the package source telemetry I added nearly a year ago, so I know from first-hand experience. Once we started getting data and tried to query it, it didn't give us useful information to make decisions.

Your concern is valid. We had discussion and asked for inputs on teams, in the end conclusion was up to me come up with what to do about this. I believe this implementation will help us answer [5 questions posted on description body](https://github.com/NuGet/NuGet.Client/pull/3674#issue-489566893) of this PR. 
Based on what we get from telemetry then we'll decide if we need more telemetry or going to have more questions need answers. We can't come up with all the questions we're trying to answer once without any preliminary basic data.
For example `ReOpenAtStart` events happens less 1.000 times in 1 million then it's pretty much self explanatory, it's very rare case and no need future action. If it happens 40.000 times in 1 million then we need to answer more questions.

I acknowledge this could be not enough, but it's good start. Please comment if you have any more input.

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs (hunk 8 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs]: Ok. I implemented your recommendation.

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs (hunk 7 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs:93]: I am not sure if it's by-design or bug. Maybe @donnie-msft  can answer this.
But considering cost of creating new object keeping same object until VS close is reasonable use of it.  'Cost' = `memory`+ `time delay` to create that object, it takes noticeable delay even with deferred Powershell load from my experience.
So if we create/destroy every time we open/close this tab the we'll see considerable time delay over time.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs (hunk 10 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: What is the relationship between .Net event and EWT events? Is EWT events does produce .Net event?
I looked that `System.Net.Http's diagnostics` but code looks snippet of code, rather than full code not sure if I can implement it this time.
Maybe static variable is easier to understand this one.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs (hunk 9 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs:337]: > VS itself already generate ton of telemetry, no need to add another telemetry for feature was not used.

It's a trade-off of usefulness vs quantity of events. We generate multiple events per restore (we really should merge them into 1 event with multiple measurements/properties, but that's a different issue), and we do one restore every build, so many per typical solution session. Hopefully we find that information useful enough to justify the quantity of data.

How valuable do you think a chart saying "for each version of VS, x% used PM UI at least once. y% used PMC at least once. z% have the 'Ask on first install which package format to use' setting enabled". I think that would be useful to know about what features customers actually care about, would be interesting for us to understand how it changes over time and across different versions of Visual Studio, and would allow us to make better decisions about what work to priotitise.  A single "NuGetFeaturesUsed" event per VS session with multiple properties could achieve this. So, I disagree that adding telemetry about features not used will lead to unreasonable telemetry volume.


> We had discussion and asked for inputs on teams, in the end conclusion was up to me come up with what to do about this.

Yes, I was part of that meeting. But I wasn't invited to a follow up meeting to discuss the proposal. As far as I can tell, you went straight to an implementation without verifying with the team that these are the things we want to collect. Maybe I'm wrong, if it was discussed and other people agree this is a good approach, then that's good.

I don't know what the pros and cons are for collecting events per VS session vs solution session, or why the decision was made to do it the way it is implemented in this PR. I don't know if I should be giving feedback only on the "high level approach", or also on the line-by-line review like PRs I expect to get merged.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs (hunk 10 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: > What is the relationship between .Net event and EWT events?

There is none. One is internal to the .NET process, the other is designed to be used for high performance monitoring of applications running on windows, with the side-effect that it's a convenient way to pass messages between components in a decoupled way.

Before making changes, I think it will be valuable to have a discussion about the telemetry we want to collect, and how you want to query it.  If you have one telemetry event with two measurements, the complexity in querying it different compared to having two events, each with a single measurement, but you have to filter which of the events are relevant to the query.

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs (hunk 7 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs:93]: VS has many tool windows. Imagine each of them using memory, even when the window is closed. This work started because the team that investigates out of memory crashes asked us to stop loading PowerShell when they thought it wasn't needed. I feel it's ironic that you're now saying it's ok for PMC to keep GUI objects in memory even when the window is closed. As for perf, customers aren't opening and closing PMC multiple times per second.

Anyway, I feel like you missed my point. My concern is from a technical correctness point of view. If it's working today as you intended by accident, then if someone in the future fixes this memory leak without knowing our telemetry depends on it, then we might start reporting telemetry differently (multiple times per session), and this could mess up our queries/reports.

Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs:93]: I agree with Andy that something is wrong with our Dispose timeline. 
If you open PMC, then close it, then open it again, it doesn't initialize. It appears that it keeps the PMC components in memory (PSHost, etc). I don't think this is correct behavior.
Example: **Developer PowerShell** (View -> Terminal)  when closed, then reopened, clearly reloads itself. I'm only guessing, but I believe this tool window is calling dispose as soon as the tool window is closed, not when VS is closed.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs (hunk 10 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs]: I combined them into 1 event emit now it would be much easier, please check .
We already held discussion and end conclusion was up to me to decide it.

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs (hunk 7 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs:93]: I am not sure what to do here, it's what it's today.  I can add comment about this concern into code.
There is no other easy way of detecting if PMC is loading automatically when you open VS. 
@donnie-msft  if you know how to detect it then please let me know.
I just discovered this detection work around by luck. I think we better make use of it for now  until there is better way of detecting it.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs (hunk 9 above)
Comment (erdembayar) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs:337]: Currently I don't know how to emit event by default when VS close from nuget, if you know please guide me for how to do it. I don't have any previous experience with `IVs` related things.
But I found where to hook when [VS close](https://github.com/NuGet/NuGet.Client/blob/d287e729bbcd6aa404c9ed26184a4d373a9e97cf/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L1113). Still I don't know how to pass telemetry info between different assemblies. 
Currently I don't see any example I can use of.
To make clear things I scheduled meeting for tomorrow.

Comment (erdembayar) [reply src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs:337]: Also one of reason I went into implementation is I need to learn first what is possible and what is not possible with VS/Nuget. 
Without any such concept it's hard to come up good questions. Now I have bit more understanding of how it works. Let's discuss tomorrow in details.

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs (hunk 7 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs:93]: > It appears that it keeps the PMC components in memory (PSHost, etc). I don't think this is correct behavior.

I think that behavior is expected right now. 
The question is whether the overhead of keeping the session alive is great. 

The UI element not getting cleaned up is definitely not ideal though.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs:93]: We recently investigated the PM UI not cleaning up the UI components, so doing the equivalent seems reasonable here.

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs (hunk 1 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs]: Yeah, that's a better phrasing. Maybe "ReopenOnStartup"?

File: src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs (hunk 7 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs:93]: Took a quick look at Developer PowerShell, and it appears to be out-of-proc. So nevermind that example.
Have we already concluded that it must remain in memory for the remainder of the VS session? Not sure if there may be other examples for just the XAML disposing.

**Comment by erdembayar** [comment]
Since current spec is under discussion let's close this PR for now.
